### PR TITLE
CSR for variable length data in embedding training

### PR DESCRIPTION
**Pull request by leezu**
## Description ##
Use sparse CSR matrices instead of masking to handle variable length inputs in embedding training.

### Essentials ###
- [x] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage
- [X] Code is well-documented

### Changes ###
- [X] CSR for variable length data in embedding training
- [X] Allow use of multiple GPUs for embedding training scripts

### Comments ###
- This is faster than the previous implementation. It also allows Hybridization.
- The refactored numba code for IO and preparing the sparse array containing words and subwords should allow further speedup is not yet parallelized. It takes around 60% of python time according to https://github.com/uber/pyflame. Depending on the batch_size, the blocking nature of nd.sparse.csr_matrix may also induce some slowdown.
- Pending verification of training results

### CODE REVIEW

**Comment by szha** [comment]
Nice. How much faster?

File: docs/examples/word_embedding/utils.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+import math
+import os
+
+import mxnet as mx
+import numpy as np
+
+import gluonnlp as nlp
+from gluonnlp.base import numba_jitclass, numba_types
+
+os.environ['MXNET_GPU_MEM_POOL_TYPE'] = 'Round'
```
Comment (szha) [inline docs/examples/word_embedding/utils.py]: this might need to happen before importing mxnet. also, would it be better to let the users set this?

File: docs/api/model.train.rst [hunk 2]
```diff
@@ -26,7 +26,7 @@ Word Embeddings
     :nosignatures:
 
     EmbeddingModel
-    SimpleEmbeddingModel
+    CSREmbeddingModel
```
Comment (leezu) [inline docs/api/model.train.rst]: May want to change to name like https://pytorch.org/docs/stable/nn.html#torch.nn.EmbeddingBag

File: docs/examples/word_embedding/utils.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            col.append(idx)
+            data.append(1 / (1 + len(self.idx_to_subwordidxs[idx])))
+            for subword in self.idx_to_subwordidxs[idx]:
+                row.append(i)
+                col.append(subword + self.offset)
+                data.append(1 / (1 + len(self.idx_to_subwordidxs[idx])))
+        return np.array(data), np.array(row), np.array(col)
+
+
+class SkipGramNet(mx.gluon.HybridBlock):
```
Comment (szha) [inline docs/examples/word_embedding/utils.py]: Should this go to `gluonnlp.model`?

File: docs/examples/word_embedding/word_embedding_training.md [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+their subword.
+Possible SubwordFunctions include mapping a word to the sequence
+of it's characters/bytes or hashes of all its ngrams.
+
+FastText models use a
+hash function to map each ngram of a word to a number in range `[0,
+num_subwords)`. We include the same hash function.
+
+### Concept of a
+SubwordFunction
```
Comment (szha) [inline docs/examples/word_embedding/word_embedding_training.md]: Fix weird line breaks.

File: docs/examples/word_embedding/word_embedding_training.md [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+```python
```
Comment (szha) [inline docs/examples/word_embedding/word_embedding_training.md]: This would become a regular python code block without being run or rendered. Follow the following guide to correct this: http://gluon-nlp.mxnet.io/how_to/contribute.html#contribute-examples

File: docs/examples/word_embedding/word_embedding_training.md [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+```
+
+We pass the dataset to the `prepare_batches` function from the `utils.py` of
+this notebook.
+
+The following actions are taken:
+- subsampling
+- negative
+samples
+- subwords in CSR
```
Comment (eric-haibin-lin) [inline docs/examples/word_embedding/word_embedding_training.md]: Many ppl are not familiar with mxnet csr. Add a link to https://mxnet.incubator.apache.org/tutorials/sparse/csr.html ?

File: docs/examples/word_embedding/utils.py (hunk 1 above)
Comment (leezu) [reply docs/examples/word_embedding/utils.py]: It shouldn't matter for the docs/examples code.The statement appears here due to copy paste. Should be resolved by sharing the utils.py code with the scripts folder once symlink is supported.

File: docs/examples/word_embedding/utils.py (hunk 3 above)
Comment (leezu) [reply docs/examples/word_embedding/utils.py]: We can add it to `gluonnlp.model.train`. This class is specific to word2vec and fastText only.

**Comment by eric-haibin-lin** [comment]
Nice work. I looked at changes in sampler, embedding.py and the notebook. Looks good except that the notebook needs to be converted again

File: docs/examples/word_embedding/utils.py (hunk 3 above)
Comment (leezu) [reply docs/examples/word_embedding/utils.py]: Let's keep it here for now and decide about moving after other embedding models are merged

**Comment by mli** [comment]
Job PR-384/19 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-384/19/index.html

**Comment by mli** [comment]
Job PR-384/20 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-384/20/index.html

File: docs/examples/word_embedding/model.py [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+    ----------
+    token_to_idx : dict
+        token_to_idx mapping of the vocabulary that this model is to be trained
+        with. token_to_idx is used for __getitem__ and __contains__. For
+        len(token_to_idx) is used during initialization to obtain the input_dim
+        of the embedding matrix.
+    output_dim : int
+        Dimension of the dense embedding.
+    batch_size : int
+        Batchsize this model will be trained with. TODO temporary until
```
Comment (szha) [inline docs/examples/word_embedding/model.py]: random_like is merged, though it's unclear whether the mxnet release can meet our timeline.

File: docs/examples/word_embedding/model.py [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+        pred_neg = F.batch_dot(emb_center, emb_negatives).squeeze()
+        mask = mask.reshape((-1, self._kwargs['num_negatives']))
+        loss_neg = (F.relu(pred_neg) + F.Activation(
+            -F.abs(pred_neg), act_type='softrelu')) * mask
+        loss_neg = loss_neg.sum(axis=1) / (mask.sum(axis=1) + 1)
+
+        return loss_pos + loss_neg
+
+
+class CBOW(Net):
```
Comment (szha) [inline docs/examples/word_embedding/model.py]: These models should reside in the API (e.g. model.train) once the interface is finalized.

File: docs/examples/word_embedding/word_embedding_training.md [hunk 9]
```diff
@@ -0,0 +1,374 @@
...
+```
+
+```{.python .input}
+pred_similarity = evaluator(words1, words2)
+sr = stats.spearmanr(pred_similarity.asnumpy(), np.array(scores))
+print('Spearman rank correlation on {} pairs of {}: {}'.format(
+    len(words1), rw.__class__.__name__, sr.correlation.round(3)))
+```
+
+# Further information
```
Comment (szha) [inline docs/examples/word_embedding/word_embedding_training.md]: Title level is incorrect, as it's showing up on the main page. http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-384/20/examples/word_embedding/word_embedding_training.html#Further-information

**Comment by szha** [comment]
http://ci.mxnet.io/blue/organizations/jenkins/gluon-nlp/detail/PR-384/21/pipeline#step-44-log-1171

**Comment by mli** [comment]
Job PR-384/23 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-384/23/index.html
